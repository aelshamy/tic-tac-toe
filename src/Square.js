import PropTypes from "prop-types";
import React from "react";

const Square = props => (
  <button className="square" onClick={props.onClick}>
    {props.value}
  </button>
);

Square.propTypes = {
  value: PropTypes.string,
  onClick: PropTypes.func
};

export default Square;
