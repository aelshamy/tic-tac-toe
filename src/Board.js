import React, { Component } from "react";
import Square from "./Square";

class Board extends Component {
  createBoard(row, col) {
    const board = [];
    let counter = 0;

    for (let i = 0; i < row; i++) {
      const columns = [];
      for (let j = 0; j < col; j++) {
        columns.push(this.renderSquare(counter++));
      }
      board.push(
        <div key={i} className="row">
          {columns}
        </div>
      );
    }

    return board;
  }

  renderSquare(idx) {
    const value = this.props && this.props.values ? this.props.values[idx] : "";
    return (
      <Square key={idx} value={value} onClick={() => this.props.onClick(idx)} />
    );
  }

  render() {
    return <div>{this.createBoard(3, 3)}</div>;
  }
}

export default Board;
