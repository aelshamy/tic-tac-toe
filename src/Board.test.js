import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import React from "react";
import { create } from "react-test-renderer";
import Board from "./Board";
import Square from "./Square";

describe("<Board/>", () => {
  beforeEach(() => {});

  test("it will match the snapshot", () => {
    const c = create(<Board />);
    expect(toJson(c)).toMatchSnapshot();
  });

  describe("createBoard()", () => {
    let board;
    beforeEach(() => {
      board = new Board();
    });
    test("should return empty board if no values provided", () => {
      const result = board.createBoard();
      expect(result).toEqual([]);
    });

    test("should return empty board if zero values provided", () => {
      const result = board.createBoard(0, 0);
      expect(result).toEqual([]);
    });

    test("should return empty board if only columns values provided", () => {
      const result = board.createBoard(null, 2);
      expect(result).toEqual([]);
    });

    test("should return only rows if rows value provided", () => {
      const result = board.createBoard(2);
      expect(result.length).toEqual(2);

      result.forEach(row => {
        expect(row.type).toEqual("div");
        expect(row.props.className).toEqual("row");
        expect(row.props.children).toEqual([]);
      });
    });

    test("should return full board if correct values provided", () => {
      spyOn(board, "renderSquare").and.returnValue(
        <span key={Math.random()} className="mock" />
      );
      const rowsValue = 3;
      const columnsValue = 3;
      const result = board.createBoard(rowsValue, columnsValue);
      // expect(Array.isArray(result)).toEqual(true);
      expect(result.length).toEqual(rowsValue);
      expect(board.renderSquare).toHaveBeenCalled();

      result.forEach(row => {
        expect(row.type).toEqual("div");
        expect(row.props.className).toEqual("row");
        expect(row.props.children.length).toEqual(columnsValue);
        row.props.children.forEach(column => {
          expect(column.type).toEqual("span");
          expect(column.props.className).toEqual("mock");
        });
      });
    });
  });

  describe("renderSquare()", () => {
    let board;
    beforeEach(() => {
      board = new Board();
    });
    test("should return Square with default value", () => {
      const result = board.renderSquare(1);
      expect(result.type).toEqual(Square);
      expect(result.key).toEqual("1");
      expect(result.props.value).toEqual("");
      expect(result.props.onClick).not.toBe(null);
    });

    test("should return Square with the correct value", () => {
      board.props = {
        values: ["first value", "second value"]
      };
      const result = board.renderSquare(1);
      expect(result.type).toEqual(Square);
      expect(result.key).toEqual("1");
      expect(result.props.value).toEqual("second value");
      expect(result.props.onClick).not.toBe(null);
    });
  });

  describe("render()", () => {
    let board;
    beforeEach(() => {
      board = new Board();
    });
    test("should call create board", () => {
      spyOn(board, "createBoard");
      board.render();
      expect(board.createBoard).toHaveBeenCalledWith(3, 3);
    });
    test("should render correctly", () => {
      const component = shallow(<Board />);
      expect(component.exists()).toBe(true);
    });
  });
});
