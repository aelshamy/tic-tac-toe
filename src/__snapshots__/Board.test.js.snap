// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Board/> it will match the snapshot 1`] = `undefined`;
